### ISSUE TITLE
Allow the server timeouts and queries per connection to be configured

### ISSUE THREAD

**Issue by rlees85** [miekg/coredns#5780]
**What would you like to be added**:

I would like to be able to configure the timeouts on the DNS server. This is NOT for the forwarding servers or the forward plugin. It is for the listener(s) on CoreDNS itself. In addition to configuring the timeouts I would also like to configure the amount of requests per connection or simply remove the limit.

CoreDNS uses: https://github.com/miekg/dns as its DNS server and this allows configuration of timeouts and requests per connection.

I would like to configure:

* ReadTimeout (https)
* WriteTimeout (https)
* IdleTimeout (https)
* ReadTimeout (tcp/tls)
* WriteTimeout (tcp/tls)
* IdleTimeout (tcp/tls)
* MaxTCPQueries (tcp/tls) - happy to just disable this by hardcoding to `-1` if the community feels this is better since I seen a comment on an issue on the miekg DNS repo that he didn't really see the need for the limit but it is left there anyway.

**Why is this needed**:

CoreDNS currently has a hard-coded 2 second read timeout for TCP/TLS connections and 5 seconds for HTTPS. This appears to be optimised for local network and/or Kubernetes environments where CoreDNS is most often used. Where CoreDNS is acting as a resolver on the public Internet this 2 second timeout can cause problems particularly on 3G/4G/5G connections where there might be lag.

**Additional details**:

I plan to prepare this pull request myself. I have never written Golang before so it should be fairly hilarious. I have already set hard-coded values myself and it builds but I would like to create something that can be merged back into official CoreDNS releases so it must be refined.

I would just like some pointers on how I should implement this? Since its not a plugin configuration it seems there is no way to put it into `Corefile` in a reasonable way. I am thinking command line parameters?

Once the PR is created hopefully the community can help a bit further around refining and tests/etc. Should hopefully only be a small one!

**Comment by chrisohaver** [comment]
Sounds reasonable to me! Though I think it would be better to do as a new plugin if we can.  This would be similar in category to other plugins such as _tls_, which do not directly process DNS requests but instead configure DNS server options.

**Comment by rlees85** [comment]
Yes after understanding plugins a bit more I agree also...

```$ ./coredns --help
Usage of ./coredns:
  -alsologtostderr
    	log to standard error as well as files
  -conf string
    	Corefile to load (default "Corefile")
  -dns.https_idle_timeout duration
    	Default DNS over HTTPS idle timeout (default 30s)
  -dns.https_read_timeout duration
    	Default DNS over HTTPS read timeout (default 5s)
  -dns.https_write_timeout duration
    	Default DNS over HTTPS write timeout (default 5s)
  -dns.port string
    	Default port (default "53")
  -dns.tcp_idle_timeout duration
    	Default TCP idle timeout (default 8s)
  -dns.tcp_max_queries int
    	Default TCP maximum queries per connection (-1 for unlimited) (default -1)
  -dns.tcp_read_timeout duration
    	Default TCP read timeout (default 2s)
  -dns.tcp_write_timeout duration
    	Default TCP write timeout (default 2s)
  -dns.tls_idle_timeout duration
    	Default DNS over TLS idle timeout (default 30s)
  -dns.tls_max_queries int
    	Default DNS over TLS maximum queries per connection (-1 for unlimited) (default -1)
  -dns.tls_read_timeout duration
    	Default DNS over TLS read timeout (default 5s)
  -dns.tls_write_timeout duration
    	Default DNS over TLS write timeout (default 5s)
  -log_backtrace_at value
    	when logging hits line file:N, emit a stack trace
  -log_dir string
    	If non-empty, write log files in this directory
  -logtostderr
    	log to standard error instead of files
  -p string
    	Default port (default "53")
  -pidfile string
    	Path to write pid file
  -plugins
    	List installed plugins
  -quiet
    	Quiet mode (no initialization output)
  -stderrthreshold value
    	logs at or above this threshold go to stderr
  -v value
    	log level for V logs
  -version
    	Show version
  -vmodule value
    	comma-separated list of pattern=N settings for file-filtered logging
```

Isn't pretty!

When I get it working under a plugin I will raise the PR. Thanks!

**Comment by kaplanben** [comment]
@rlees85, we are facing the same case and your PR saved weeks of work for me so thank you!

**Comment by Tantalor93** [comment]
thanks! A new plugin providing the desired functionality was merged

**Comment by rlees85** [comment]
@kaplanben nice to hear.... curious your use case? Originally it was to try and work with Fritz Box routers but it didn't help. It stopped me getting private DNS errors on Android a lot on 3/4/5g though hence why pressing ahead. Desperately needed to understand Go a bit more as well

**Comment by kaplanben** [comment]
@rlees85, our case is about timeouts over TCP when the backend (authoritative) server has high latency (far away). We noticed that when using load test tools like `DNSperf/Resperf` in this scenario most of the request timeout when using **a single client**. If we increase the `act as n clients` argument to `> 10` there are zero timeouts. Also, when doing the same test directly to the remote backend (authoritative) we get zero timeouts as well.

I was sure that this is related to `MaxTCPQueries`, but after testing this with your PR we saw that the issue persists.
Later, I switch the coreDNS with DNSDist and saw the same behavior so this is not a case specifically affecting CoreDNS but for the life of me we can't narrow it down (similar cases are [PDNS repo](https://github.com/PowerDNS/pdns/issues/623) and [DNSperf Repo](https://github.com/DNS-OARC/dnsperf/issues/63) both closed with no real explanation).
